server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java

```diff
@@ -77,6 +77,16 @@ public final long getUsageCount() {
 
     @Override
     public final void handleRequest(RestRequest request, RestChannel channel, NodeClient client) throws Exception {
+        // check if the query has any parameters that are not in the supported set (if declared)
+        var supported = supportedQueryParameters();
+        if (supported != null) {
+            Set<String> unsupported = new HashSet<>(request.params().keySet());
+            unsupported.removeAll(supported);
+            if (unsupported.isEmpty() == false) {
+                throw new IllegalArgumentException(unrecognized(request, unsupported, supported, "parameter"));
```

Comment:
It could be problematic to do the check here, in `handleRequest`, against the query parameters that are specified.
To be "safer" and more lenient, IMO we could do parameter at "read time", only when and if we try to access them (in prepareRequest or elsewhere).
Suppose you add a query parameter to an API; if you are building and using the same API call against an older version, the current behaviour is that the new unrecognized parameter is ignored, so you can change your calls and add things without breaking.
True, once we have the capabilities API you shouldn't, as you could build your query string based on capabilities, but if the client is old or basic that would not work.

It might be something that we want, to be more strict, but we have to be very aware - and to me it seems a bit unnecessary.

Reply:
Should we skip the validation of unconsumed params if we already validated the supported ones?

Reply:
I think we need both - checking that all the params specified are supported (call check for the user) and that the handler has done something with the parameters that were specified (handler implementation check for the developer)

Reply:
> I think the only thing missing is to always consider all general params (`responseParams(request.getRestApiVersion())`) to be supported as well. And treat them similarly when doing the capabilities check.

I don't think we should do this; the two methods have got different functionality. `RestHandler.supportedQueryParameters()` is used to declare the complete set of query parameters supported for callers (once the  handler has been modified to specify this), and `BaseRestHandler.responseParams` is implementation functionality around parameter consumption validation, which is part of the developer implementation.

Reply:
I see you introduced REST api major version, will `supportedQueryParameters` depend on that?

Reply:
hmm, yes, it is dependent on that, as well as features (or whatever they're called).

Reply:
Ok, great 👍